fe_utils/solvers/mastery.py

```diff
@@ -2,9 +2,133 @@
+    the function space in which to solve and the right hand side
+    function."""
+
+    # Create an appropriate (complete) quadrature rule.
+
+    # Tabulate the basis functions and their gradients at the quadrature points.
+
+    # Create the left hand side matrix and right hand side vector.
+    # This creates a sparse matrix because creating a dense one may
+    # well run your machine out of memory!
+
+    A = sp.lil_matrix((fs1.node_count, fs1.node_count))
+    B = sp.lil_matrix((fs2.node_count, fs1.node_count))
+    rhs = np.zeros(fs1.node_count + fs2.node_count)
+    l = rhs[:fs1.node_count]
+
+    # Now loop over all the cells and assemble A and l
+    fe1 = fs1.element
+    fe2 = fs2.element
+    d = fs1.element.cell.dim
+    mesh = fs1.mesh
+
+    # Create a quadrature rule which is exact for (f1-f2)**2.
+    Q = gauss_quadrature(fe1.cell, 2*max(fe1.degree, fe2.degree))
+    # Q2 = gauss_quadrature(fe2.cell, 2*max(fe1.degree, fe2.degree))
+
+    # Evaluate the local basis functions at the quadrature points.
+    phi1 = fe1.tabulate(Q.points)
+    phi1_grad = fe1.tabulate(Q.points, grad=True)
+    phi2 = fe2.tabulate(Q.points)
+    phi2_grad = fe2.tabulate(Q.points, grad=True)
+
+    val = 0.
+    for c in range(mesh.entity_counts[-1]):
+        # Find the appropriate global node numbers for this cell.
+        nodes1 = fs1.cell_nodes[c, :]
+        nodes2 = fs2.cell_nodes[c, :]
+
+        # Compute the change of coordinates.
+        J = mesh.jacobian(c)
+        detJ = np.abs(np.linalg.det(J))
+        # b = basis, d = dimension, q= quadrature 
+        f_quadrature = np.einsum("b, bdq -> dq", f.values[nodes1], phi1.T)
+        l[nodes1] += np.einsum("bdq, dq -> b", phi1.T, f_quadrature * Q.weights) * detJ 
+
+        # # i = gradient component
+        # # ji dimensions of jacobian
+        # J_phi = np.einsum("ji,bdqi -> bdqi", np.linalg.inv(J.T), phi1_grad)
+        # div = np.sum(np.array([J_phi[:, j, :, j] for j in range(d)]), axis=0)
+        # # c = basis of other space
+        # B[np.ix_(nodes2, nodes1)] += np.einsum("q, qc, qb -> cb", Q.weights, phi2, div) * detJ 
+
+        div = np.array([phi1_grad[:, j, :, j] for j in range(d)])
+        Jdiv = np.einsum("ij,jkm -> km", np.linalg.inv(J.T), div)
+        B[np.ix_(nodes2, nodes1)] += np.einsum("i, ij, ik -> jk", Q.weights, phi2, Jdiv) * detJ 
+
+        # # phi_term = np.einsum("iq, jq -> ijq", phi1.T, phi1.T)
+        
+        # half_grad_term = np.einsum("li, ijkm -> ljkm", np.linalg.inv(J.T), phi1_grad.T)
+        half_grad_term = np.einsum("ji,bdqi -> bdqi", np.linalg.inv(J.T), phi1_grad)
+        half_grad_termT = np.einsum("bdqi -> biqd", half_grad_term)
+        # half_grad_termT = np.einsum("ljkm -> kjlm", half_grad_term)
+        grad_term = (1/2)*(half_grad_term + half_grad_termT)
+        
+        # twice_sum = np.einsum("ijkm -> jm", grad_term * grad_term)
+        twice_sum = np.einsum("biqd ->bq ", grad_term * grad_term)
```

Comment:
I am also not convinced that this is the right einsum for the inner product of grad_term with itself. It's certainly not the obvious way of writing it.

Reply:
I think I'm rather confused about what is meant to happen here- what exactly does the : in the notes mean?

Reply:
: is the inner product. So: 
![formula](https://render.githubusercontent.com/render/math?math=\nabla%20u%20:%20\nabla%20u%20=%20\sum_{\alpha,\beta}%20\frac{\partial%20u_\alpha}{\partial%20x_\beta}\frac{\partial%20u_\alpha}{\partial%20x_\beta})

Reply:
alright thank you!